jbpm-event-emitters/jbpm-event-emitters-elasticsearch/src/main/java/org/jbpm/event/emitters/elasticsearch/ElasticSearchEventEmitter.java

```diff
@@ -70,64 +64,69 @@
+            }
+        }
+        return factory != null ? factory : new DefaultESInstanceViewTransformerFactory();
+    }
+    
     private ObjectMapper mapper = new ObjectMapper();
+    
+    private ESInstanceViewTransformerFactory factory = getFactory();
+   
 
     private ExecutorService executor;
     
     private CloseableHttpClient httpclient;
 
     public ElasticSearchEventEmitter() {
-        mapper.setDateFormat(new SimpleDateFormat(dateFormatStr));
-        mapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
-        mapper.configure(MapperFeature.PROPAGATE_TRANSIENT_MARKER, true);
-        
+        factory.configureObjectMapper(mapper);
         executor = buildExecutorService();
         httpclient = buildClient();
     }
+    
 
+    @Override
     public void deliver(Collection<InstanceView<?>> data) {
         // no-op
 
     }
 
+    @Override
     public void apply(Collection<InstanceView<?>> data) {
         if (data.isEmpty()) {
             return;
         }
 
         executor.execute(() -> {
             StringBuilder content = new StringBuilder();
-
+            Set<String> visitedIds = new HashSet<>();
             for (InstanceView<?> view : data) {
                 try {
-                    String json = mapper.writeValueAsString(view);
-
-                    String index = "jbpm";
-                    String type = "unknown";
-                    String id = "";
-                    if (view instanceof ProcessInstanceView) {
-                        index = "processes";
-                        type = "process";
-                        id = ((ProcessInstanceView) view).getCompositeId();
-                    } else if (view instanceof TaskInstanceView) {
-                        index = "tasks";
-                        type = "task";
-                        id = ((TaskInstanceView) view).getCompositeId();
-                    } else if (view instanceof CaseInstanceView) {
-                        index = "cases";
-                        type = "case";
-                        id = ((CaseInstanceView) view).getCompositeId();
+                    ESInstanceViewTransformer instanceView = factory.getInstanceViewTransfomer(view);
+                    if (instanceView == null) {
+                        logger.warn("Unsupported view {} ", view.getClass());
+                        continue;
                     }
-
-                    content.append("{ \"index\" : { \"_index\" : \"" + index + "\", \"_type\" : \"" + type + "\", \"_id\" : \"" + id + "\" } }\n");
-                    content.append(json);
-                    content.append("\n");
-
+                    ESRequest request = visitedIds.add(view.getCompositeId()) ? instanceView.index(view) : instanceView
+                            .update(view);
+                    content.append(String.format(
+                            "{ \"%s\" : { \"_index\" : \"%s\", \"_type\" : \"%s\", \"_id\" : \"%s\" } }%n%s%n", request
```

Comment:
As far as I know the `_type` field was [deprecated on Elasticsearch 6](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-type-field.html#mapping-type-field). Actually it's not used by the `_bulk` API (see [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html#bulk-api-request-body)).

Is it kept for backwards compatibility?

Reply:
TBH I did not revisit that part of the emitter, but yes, it is better to keep it for backward compatibility